test/integration/process-statement-message.test.js

```diff
@@ -0,0 +1,70 @@
+const { BlobServiceClient } = require('@azure/storage-blob')
+const config = require('../../app/config/storage')
+const db = require('../../app/data')
+const mockStatement = require('../mocks/statement-data')
+const processStatementMessage = require('../../app/messaging/process-statement-message')
+
+const FILE_NAME = 'FFC_PaymentStatement_SFI_2022_1234567890_2022080515301012.pdf'
+
+let blobServiceClient
+let container
+let receiver
+let message
+
+describe('generate statements', () => {
+  beforeEach(async () => {
+    jest.useFakeTimers().setSystemTime(new Date(2022, 7, 5, 15, 30, 10, 120))
+    blobServiceClient = BlobServiceClient.fromConnectionString(config.connectionStr)
+    container = blobServiceClient.getContainerClient(config.container)
+    await container.deleteIfExists()
+    await container.createIfNotExists()
+    await db.sequelize.truncate({ cascade: true })
+
+    receiver = {
+      completeMessage: jest.fn()
+    }
+
+    message = {
+      body: mockStatement
+    }
+  })
+
+  afterAll(async () => {
+    await db.sequelize.truncate({ cascade: true })
+    await db.sequelize.close()
+  })
+
+  test('publishes PDF statement to blob storage outbound location', async () => {
+    await processStatementMessage(message, receiver)
+    const fileList = []
+    for await (const item of container.listBlobsFlat({ prefix: config.archiveFolder })) {
+      fileList.push(item.name)
+    }
+    expect(fileList.filter(x => x === `${config.folder}/${FILE_NAME}`).length).toBe(1)
+  })
+
+  test('saves log entry', async () => {
+    await processStatementMessage(message, receiver)
+    const log = await db.generation.findOne({ where: { filename: `${FILE_NAME}` } })
+    expect(log).not.toBeNull()
+    expect(log.statementData).toStrictEqual(mockStatement)
+    expect(log.dateGenerated).toStrictEqual(new Date(2022, 7, 5, 15, 30, 10, 120))
+  })
+
+  test('saves log entry with statement data', async () => {
+    await processStatementMessage(message, receiver)
+    const log = await db.generation.findOne({ where: { filename: `${FILE_NAME}` } })
+    expect(log.statementData).toStrictEqual(mockStatement)
+  })
+
+  test('saves log entry with generation date', async () => {
+    await processStatementMessage(message, receiver)
+    const log = await db.generation.findOne({ where: { filename: `${FILE_NAME}` } })
+    expect(log.dateGenerated).toStrictEqual(new Date(2022, 7, 5, 15, 30, 10, 120))
+  })
+
+  test('completes message', async () => {
+    await processStatementMessage(message, receiver)
+    expect(receiver.completeMessage).toHaveBeenCalled()
```

Comment:
no integration test possible for invalid message becuz we dnt check/connect to ServiceBus, i.e. check that cleared down Topic has 0 messages in both Queue and DLQ, then run function w/ good request (see Queue have 1 message, verify its the one we sent) or bad request (see DLQ [after a while due to retry] have 1 message, verify its the one we sent)

We dnt do this because we have no local Azure ServiceBus emulator, right?

Reply:
Yes that's the main reason we don't.  If had an emulator we would do exactly as you suggest, similar to what we do with Postgres when saving records.  Means our tests would give us much more confidence as they're using real protocols and integrations.

It is possible to do without emulation and we have done it in the demo service, but it involves pain.

The Jenkins pipeline will provision a full set of topics/subscriptions based on the contents of `provision.azure.yaml` unique to that build.  So every CI build does have a set of collision free queues with the intention they can be used for integration tests.  Once the build is complete those queues are automatically deleted.

But if you've got multiple tests writing messages to those queues, because of the async nature of messaging you can't necessarily guarantee those messages will be pulled off the queue in the right order.  So you need to think about how your tests wait for the right message for the specific test.  You also need to run test not in parallel (as we do anyway so we can managed connections consistently)  and have them setup to clear the queues before and after each test execution.  So you end up with more service bus processing code in your tests to do this as there isn't a nice `queue.clear()` command.

Ideally you'd have a unique set of queues per build per test, but that's a lot of work in Jenkins we haven't done to solve that problem.  There is a service bus sdk that supports creation and deletion of entities from Node.js so another solution would be for tests to manage their dependencies themselves.  But what if a test fails in CI, the clean up steps don't run and we end up with a treasure trove of orphan queues over time without further effort to make sure they're always cleaned up.

The other problem it gives us is local development outside of CI.  We have one set of queues per developer so they would share the same queues as local running of tests as they would for running the app.  Again the tests could empty the queues before they run each one, but what if I've spent ages loading my queue with messages for local dev.  I don't want the tests to purge all that.

Finally, we've been using service bus long enough that we've got high confidence that if we call `complete`, `deadLetter` , `abandon` etc then service bus will do as we ask.  So there is risk associated with not testing it, but the risk is so low it's just not worth all the pain to do it.

One positive I did notice though from Microsoft is this comment on the (long) open GitHub issue for local emulation.

![image](https://user-images.githubusercontent.com/42574977/185118739-0b4a584f-e4f5-46f1-95e3-5242503d143c.png)

https://github.com/Azure/azure-service-bus/issues/223

Just because it's on a backlog doesn't mean it will get done but it's the most positive sign I've seen in years!